desc/objectives/utils.py

```diff
@@ -371,6 +372,43 @@ def align_jacobian(Fx, objective_f, objective_g):
     return A.T
 
 
+def jax_softmax(arr, alpha):
+    """JAX softmax implementation.
+
+    Inspired by https://www.johndcook.com/blog/2010/01/13/soft-maximum/
+    and https://www.johndcook.com/blog/2010/01/20/how-to-compute-the-soft-maximum/
+
+    Parameters
+    ----------
+    arr: ndarray, the array which we would like to apply the softmax function to.
+    alpha: float, the parameter smoothly transitioning the function to a hardmax.
+        as alpha increases, the value returned will come closer and closer to
+        max(arr).
+
+    Returns
+    -------
+    softmax: float, the soft-maximum of the array.
+    """
+    return logsumexp(arr * alpha) / alpha
+
+
+def jax_softmin(arr, alpha):
+    """JAX softmin implementation, by taking negative of softmax(-arr).
+
+    Parameters
+    ----------
+    arr: ndarray, the array which we would like to apply the softmin function to.
+    alpha: float, the parameter smoothly transitioning the function to a hardmin.
+        as alpha increases, the value returned will come closer and closer to
+        min(arr).
+
+    Returns
+    -------
+    softmin: float, the soft-minimum of the array.
+    """
+    return -jax_softmax(-arr, alpha)
```

Comment:
@f0uriest if I were to add a line here that checked that alpha*arr was > 1, if not it multiplied by a factor to make it so, then divided by it, would that end up messing with the optimization? it is multiplication and division by the same quantity so it should be fine, though when it gets jitted I am not sure if that will freeze the if statement logic or not

like

```python
if jnp.any(jnp.abs(alpha*arr)<1):
    factor = jnp.min(jnp.abs(alpha*arr))
else:
    factor = 1
return -jax_softmax(-arr/factor,alpha)*factor
```

to avoid cases where it is inaccurate bc array elements are <1

Reply:
or maybe
```python
from jax import lax

arr_times_alpha = alpha*arr
min_val = jnp.min(jnp.abs(arr_times_alpha))
return lax.cond(jnp.any(jnp.abs(arr_times_alpha)<1),
                lambda arr_times_alpha: logsumexp(arr_times_alpha / min_val ) / alpha * min_val,
                lambda arr_times_alpha: logsumexp(arr_times_alpha) / alpha,
                arr_times_alpha)
 ```
for the softmax function. This does add some overhead (it is ~twice as slow as not having this conditional), and I am not even sure how much this helps with this issue according to the tests I am running right now...

Reply:
This should work fine with jit, but I am not sure how this affects the derivative since it technically is using the min function...